backend/ntlm.js

```diff
@@ -0,0 +1,463 @@
+  buf.writeUInt32LE(BODY_LENGTH + domainNameBytes.length + usernameBytes.length + workstationBytes.length + lmChallengeResponse.length, pos); pos += 4; // NtChallengeResponseOffset
+
+  buf.writeUInt16LE(domainNameBytes.length, pos); pos += 2; // DomainNameLen
+  buf.writeUInt16LE(domainNameBytes.length, pos); pos += 2; // DomainNameMaxLen
+  buf.writeUInt32LE(BODY_LENGTH, pos); pos += 4;            // DomainNameOffset
+
+  buf.writeUInt16LE(usernameBytes.length, pos); pos += 2; // UserNameLen
+  buf.writeUInt16LE(usernameBytes.length, pos); pos += 2; // UserNameMaxLen
+  buf.writeUInt32LE(BODY_LENGTH + domainNameBytes.length, pos); pos += 4; // UserNameOffset
+
+  buf.writeUInt16LE(workstationBytes.length, pos); pos += 2; // WorkstationLen
+  buf.writeUInt16LE(workstationBytes.length, pos); pos += 2; // WorkstationMaxLen
+  buf.writeUInt32LE(BODY_LENGTH + domainNameBytes.length + usernameBytes.length, pos); pos += 4; // WorkstationOffset
+
+  buf.writeUInt16LE(encryptedRandomSessionKeyBytes.length, pos); pos += 2; // EncryptedRandomSessionKeyLen
+  buf.writeUInt16LE(encryptedRandomSessionKeyBytes.length, pos); pos += 2; // EncryptedRandomSessionKeyMaxLen
+  buf.writeUInt32LE(BODY_LENGTH + domainNameBytes.length + usernameBytes.length + workstationBytes.length + lmChallengeResponse.length + ntChallengeResponse.length, pos); pos += 4; // EncryptedRandomSessionKeyOffset
+
+  // Fix #98
+  var flagsToWrite = isUnicode
+    ? typeflags.NTLM_TYPE2_FLAGS
+    : typeflags.NTLM_TYPE2_FLAGS - flags.NTLM_NegotiateUnicode;
+  buf.writeUInt32LE(flagsToWrite , pos); pos += 4; // NegotiateFlags
+
+  buf.writeUInt8(5, pos); pos++; // ProductMajorVersion
+  buf.writeUInt8(1, pos); pos++; // ProductMinorVersion
+  buf.writeUInt16LE(2600, pos); pos += 2; // ProductBuild
+  buf.writeUInt8(0, pos); pos++; // VersionReserved1
+  buf.writeUInt8(0, pos); pos++; // VersionReserved2
+  buf.writeUInt8(0, pos); pos++; // VersionReserved3
+  buf.writeUInt8(15, pos); pos++; // NTLMRevisionCurrent
+
+  domainNameBytes.copy(buf, pos); pos += domainNameBytes.length;
+  usernameBytes.copy(buf, pos); pos += usernameBytes.length;
+  workstationBytes.copy(buf, pos); pos += workstationBytes.length;
+  lmChallengeResponse.copy(buf, pos); pos += lmChallengeResponse.length;
+  ntChallengeResponse.copy(buf, pos); pos += ntChallengeResponse.length;
+  encryptedRandomSessionKeyBytes.copy(buf, pos); pos += encryptedRandomSessionKeyBytes.length;
+
+  return "NTLM " + buf.toString("base64");
+}
+
+function create_LM_hashed_password_v1(password) {
+  // fix the password length to 14 bytes
+  password = password.toUpperCase();
+  var passwordBytes = Buffer.from(password, "ascii");
+
+  var passwordBytesPadded = Buffer.alloc(14);
+  passwordBytesPadded.fill("\0");
+  var sourceEnd = 14;
+  if (passwordBytes.length < 14) {
+    sourceEnd = passwordBytes.length;
+  }
+  passwordBytes.copy(passwordBytesPadded, 0, 0, sourceEnd);
```

Comment:
```suggestion
  passwordBytes.copy(passwordBytesPadded);
```

Reply:
@NeilRashbrook I've commited those changes. But...
Why do you think it's unnecessary to fill the buffer with 0s?

Reply:
`new Buffer(size)` originally gave you uninitialised memory. This was changed in Node 8 to always zero-initialise the buffer. However, it was also deprecated in Node 6; in some versions of Node 7 it would always log a warning to the console, although since Node 10 it only does that for your main code, not in a library. If this code predates Node 8 then they could have converted from `new Buffer` to `Buffer.alloc` and possibly not realised that this gives you zero-initialised memory?